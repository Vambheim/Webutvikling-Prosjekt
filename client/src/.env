# Måte å tilby kryptert API-nøkkel (gir tryggere kommunikasjon mot nettet) 
REACT_APP_API_KEY=5b7869e75f75481a8f5005bc7bffd3a8 
