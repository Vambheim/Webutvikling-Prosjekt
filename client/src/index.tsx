import ReactDOM from 'react-dom';
import * as React from 'react';
import { Component } from 'react-simplified';
import { HashRouter, Route } from 'react-router-dom';
import { Alert, Card } from './widgets';
import {
  RecipeAdd,
  RecipeList,
  ShoppingList,
  TaskDetails,
  TaskEdit,
  TaskNew,
} from './recipe-components';
import Container from 'react-bootstrap/Container';
import Nav from 'react-bootstrap/Nav';
import Navbar from 'react-bootstrap/Navbar';
import NavDropdown from 'react-bootstrap/NavDropdown';
// import Card from 'react-bootstrap/Card';

class Menu extends Component {
  render() {
    return (
      <Navbar bg="light" expand="lg">
        <Container>
          <Navbar.Brand href="/" className="ms-auto">
            Java matapplikasjon
          </Navbar.Brand>
          <Navbar.Toggle aria-controls="basic-navbar-nav" />
          <Navbar.Collapse id="basic-navbar-nav">
            <Nav className="ms-auto">
              <Nav.Link href="/#/recipes">Oppskrifter</Nav.Link>
              <Nav.Link href="/#/recipes/add">Legg til oppskrift</Nav.Link>
              <Nav.Link href="/#/recipes/cart">Handleliste</Nav.Link>
              <NavDropdown title="Innstillinger" id="basic-nav-dropdown">
                <NavDropdown.Item href="/#/recipes/cart">Profil</NavDropdown.Item>
                <NavDropdown.Item href="#action/3.2">MÃ¸rk modus</NavDropdown.Item>
                <NavDropdown.Divider />
                <NavDropdown.Item href="#action/3.4">Overrask meg</NavDropdown.Item>
              </NavDropdown>
            </Nav>
          </Navbar.Collapse>
        </Container>
      </Navbar>
    );
  }
}

class Home extends Component {
  render() {
    return <Card title="Welcome">This is your favourite food recipe app</Card>;
  }
}

ReactDOM.render(
  <HashRouter>
    <div>
      <Alert />
      <Menu />
      <Route exact path="/" component={Home} />
      <Route exact path="/recipes" component={RecipeList} />
      <Route exact path="/recipes/add" component={RecipeAdd} />
      <Route exact path="/recipes/cart" component={ShoppingList} />
      <Route exact path="/tasks/:id(\d+)" component={TaskDetails} /> {/* id must be number */}
      <Route exact path="/tasks/:id(\d+)/edit" component={TaskEdit} /> {/* id must be number */}
      <Route exact path="/tasks/new" component={TaskNew} />
    </div>
  </HashRouter>,
  document.getElementById('root')
);
