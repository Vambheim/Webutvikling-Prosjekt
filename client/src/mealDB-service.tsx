import axios from 'axios';

axios.defaults.baseURL = 'https://www.themealdb.com/api/json/v1/1';

export type MealDBRecipe = {
  mealDBId: number;
  name: string;
  category: string;
  country: string;
};

class MealDBService {
  /**
   * Get task with given id.
   */
  get(mealDBId: number) {
    return axios.get<MealDBRecipe>('/lookup.php?i=' + mealDBId).then((response) => response.data);
  }

  /**
   * Get all tasks.
   */
  // getAll() {
  //   return axios.get<Recipe[]>('/recipes').then((response) => response.data);
  // }

  /**
   * Create new task having the given title.
   *
   * Resolves the newly created task id.
   */
  // create(title: string) {
  //   return axios
  //     .post<{ id: number }>('/tasks', { title: title })
  //     .then((response) => response.data.id);
  // }
}

const mealDBService = new MealDBService();
export default mealDBService;
