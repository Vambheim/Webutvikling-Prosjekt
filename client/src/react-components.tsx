import * as React from 'react';
import { Component } from 'react-simplified';
import { NavLink } from 'react-router-dom';
import taskService, { Task } from './recipe-service';
import { createHashHistory } from 'history';
import {
  Card,
  Container,
  Row,
  Col,
  Carousel,
  Button,
  Form,
  Navbar,
  Nav,
  NavDropdown,
  Accordion,
} from 'react-bootstrap';
import { Column } from './widgets';

// Use history.push(...) to programmatically change path, for instance after successfully saving a student
const history = createHashHistory();

export class Home extends Component {
  render() {
    return (
      <Container style={{ flex: 'column' }}>
        <Card style={{ border: 'none', textAlign: 'center', fontFamily: 'Century Gothic' }}>
          <Card.Body>
            <Card style={{ border: 'none', textAlign: 'center' }}>
              <Card.Body>
                <Carousel>
                  <Carousel.Item interval={1500}>
                    <img
                      className="d-block w-100"
                      src="https://s.tihlde.org/burger"
                      alt="First slide"
                      width={'900px'}
                    />
                    <Carousel.Caption>
                      <h3>Burger til under 100kr</h3>
                      <p>Nulla vitae elit libero, a pharetra augue mollis interdum.</p>
                    </Carousel.Caption>
                  </Carousel.Item>
                  <Carousel.Item interval={500}>
                    <img
                      className="d-block w-100"
                      src="https://s.tihlde.org/pannekake"
                      alt="Second slide"
                    />
                    <Carousel.Caption>
                      <h3>Dessert til enhver anledning</h3>
                      <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p>
                    </Carousel.Caption>
                  </Carousel.Item>
                  <Carousel.Item>
                    <img
                      className="d-block w-100"
                      src="https://s.tihlde.org/brownie"
                      alt="Third slide"
                    />
                    <Carousel.Caption>
                      <h3>Frokost eller middag?</h3>
                      <p>Praesent commodo cursus magna, vel scelerisque nisl consectetur.</p>
                    </Carousel.Caption>
                  </Carousel.Item>
                </Carousel>
              </Card.Body>
              <Card.Title>
                {'Anbefalinger:'}
                <Card.Body style={{ backgroundColor: 'rgb(169, 110, 41)' }}>
                  <Row style={{ justifyContent: 'center' }}>
                    <Card
                      style={{
                        width: '18rem',
                        marginRight: '1rem',
                        backgroundColor: 'rgb(255, 235, 205)',
                      }}
                    >
                      <Card.Img
                        variant="top"
                        src="https://s.tihlde.org/skinke"
                        style={{ marginTop: '10px' }}
                      />
                      <Card.Body>
                        <Card.Title>Card Title</Card.Title>
                        <Card.Text>
                          Some quick example text to build on the card title and make up the bulk of
                          the card's content.
                        </Card.Text>
                        <Button variant="primary">Go somewhere</Button>
                      </Card.Body>
                    </Card>
                    <Card
                      style={{
                        width: '18rem',
                        marginRight: '1rem',
                        backgroundColor: 'rgb(255, 235, 205)',
                      }}
                    >
                      <Card.Img
                        variant="top"
                        src="https://s.tihlde.org/suppe"
                        style={{ marginTop: '10px' }}
                      />
                      <Card.Body>
                        <Card.Title>Card Title</Card.Title>
                        <Card.Text>
                          Some quick example text to build on the card title and make up the bulk of
                          the card's content.
                        </Card.Text>
                        <Button variant="primary">Go somewhere</Button>
                      </Card.Body>
                    </Card>
                    <Card
                      style={{
                        width: '18rem',
                        marginRight: '1rem',
                        backgroundColor: 'rgb(255, 235, 205)',
                      }}
                    >
                      <Card.Img
                        variant="top"
                        src="https://s.tihlde.org/suppe"
                        style={{ marginTop: '10px' }}
                      />
                      <Card.Body>
                        <Card.Title>Card Title</Card.Title>
                        <Card.Text>
                          Some quick example text to build on the card title and make up the bulk of
                          the card's content.
                        </Card.Text>
                        <Button variant="primary">Go somewhere</Button>
                      </Card.Body>
                    </Card>
                    <Card
                      style={{
                        width: '18rem',
                        marginRight: '1rem',
                        backgroundColor: 'rgb(255, 235, 205)',
                      }}
                    >
                      <Card.Img
                        variant="top"
                        src="https://s.tihlde.org/suppe"
                        style={{ marginTop: '10px' }}
                      />
                      <Card.Body>
                        <Card.Title>Card Title</Card.Title>
                        <Card.Text>
                          Some quick example text to build on the card title and make up the bulk of
                          the card's content.
                        </Card.Text>
                        <Button variant="primary">Go somewhere</Button>
                      </Card.Body>
                    </Card>
                  </Row>
                </Card.Body>
              </Card.Title>
            </Card>
          </Card.Body>
        </Card>
      </Container>
    );
  }
}

export class RecipeList extends Component {
  tasks: Task[] = [];
  countries: string[] = ['Norway', 'China']; // Midlertidig løsning frem til api er hentet
  country: string = '';
  categories: string[] = ['Fish', 'Tapas']; // Midlertidig løsning frem til api er hentet
  category: string = '';
  ingredients: string[] = ['Milk', 'Chilli']; // Midlertidig løsning frem til api er hentet
  ingredient: string = '';

  render() {
    return (
      <div>
        <Container fluid>
          <Navbar expand="lg" variant="light" bg="">
            <Container style={{ display: 'flex', justifyContent: 'normal' }}>
              <Navbar.Brand href="">Filters:</Navbar.Brand>
              <Navbar.Toggle aria-controls="basic-navbar-nav" />
              <Navbar.Collapse id="basic-navbar-nav">
                <Nav className="m-auto" style={{ width: '50%' }}>
                  <Form.Select
                    value={this.country}
                    onChange={(event) => (this.country = event.currentTarget.value)}
                    style={{ border: 'none', marginRight: '10px' }}
                  >
                    <option key="blankChoice" hidden>
                      {'Velg land:'}
                    </option>

                    {this.countries.map((c) => (
                      <option key={c} value={c}>
                        {c}
                      </option>
                    ))}
                  </Form.Select>

                  <Form.Select
                    value={this.category}
                    onChange={(event) => (this.category = event.currentTarget.value)}
                    style={{ border: 'none', marginRight: '10px' }}
                  >
                    <option key="blankChoice" hidden>
                      {'Velg Kategori:'}
                    </option>

                    {this.categories.map((c) => (
                      <option key={c} value={c}>
                        {c}
                      </option>
                    ))}
                  </Form.Select>
                  <Form.Select
                    value={this.ingredient}
                    onChange={(event) => (this.ingredient = event.currentTarget.value)}
                    style={{ border: 'none', marginRight: '10px' }}
                  >
                    <option key="blankChoice" hidden>
                      {'Velg ingrediens'}
                    </option>
                    {this.ingredients.map((ing) => (
                      <option key={ing} value={ing}>
                        {ing}
                      </option>
                    ))}
                  </Form.Select>
                </Nav>
              </Navbar.Collapse>
            </Container>
          </Navbar>
          <Card>
            <Card.Title>
              <Card.Body>
                <Row xs={2} md={4} className="g-4">
                  {Array.from({ length: 16 }).map((_, idx) => (
                    <Col>
                      <Card>
                        <Card.Img variant="top" src="https://s.tihlde.org/brownie" />
                        <Card.Body>
                          <Card.Title>Card title</Card.Title>
                          <Card.Text>
                            This is a longer card with supporting text below as a natural lead-in to
                            additional content. This content is a little bit longer.
                          </Card.Text>
                        </Card.Body>
                      </Card>
                    </Col>
                  ))}
                </Row>
              </Card.Body>
            </Card.Title>
          </Card>
        </Container>
      </div>
    );
  }
}

export class RecipeAdd extends Component {
  categories: string[] = ['Fish', 'Tapas']; // Midlertidig løsning frem til api er hentet
  category: string = '';
  recipeName: string = '';
  description: string = '';
  country: string = '';
  ingredients: string = '';
  ingredient: string[] = ['Carrot', 'Potato', 'Chicken'];

  render() {
    return (
      <Container style={{}}>
        <Card
          style={{
            border: 'none',
            textAlign: 'center',
            marginTop: '20px',
          }}
        >
          <Card.Title>New Recipe</Card.Title>
          <Card.Body>
            <Card.Text>Want to add a new recipe to our database? Add it below: </Card.Text>
          </Card.Body>
        </Card>
        <Row>
          <Col md={2}>
            <Form>
              <Form.Group className="mb-3" controlId="formBasicname">
                <Form.Label>Name</Form.Label>
                <Form.Control
                  type="Name"
                  placeholder="Enter name"
                  value={this.recipeName}
                  onChange={(event) => (this.recipeName = event.currentTarget.value)}
                />
                <Form.Text className="text-muted"></Form.Text>
              </Form.Group>
              <Form.Group className="mb-3" controlId="formBasicname">
                <Form.Label>Description</Form.Label>
                <Form.Control
                  type="desc"
                  placeholder="Enter description"
                  value={this.description}
                  onChange={(event) => (this.description = event.currentTarget.value)}
                />
                <Form.Text className="text-muted"></Form.Text>
              </Form.Group>
              <Form.Group className="mb-3" controlId="formBasicname">
                <Form.Label>Country</Form.Label>
                <Form.Control
                  type="Country"
                  placeholder="Enter country"
                  value={this.country}
                  onChange={(event) => (this.country = event.currentTarget.value)}
                />
                <Form.Text className="text-muted"></Form.Text>
              </Form.Group>
              Kategori
              <Form.Select
                value={this.category}
                onChange={(event) => (this.category = event.currentTarget.value)}
              >
                <option key="Category" hidden></option>
                {this.categories.map((c) => (
                  <option key={c} value={c}>
                    {c}
                  </option>
                ))}
              </Form.Select>
              <Form.Group className="collapse" controlId="formBasicIngredient" hidden>
                <Form.Label>Ingredients</Form.Label>
                <Form.Control
                  type="Ingredient"
                  placeholder="Enter Ingredient"
                  value={this.ingredients}
                  onChange={(event) => (this.ingredients = event.currentTarget.value)}
                />
                <Form.Text className="text-muted"></Form.Text>
              </Form.Group>
              <Accordion defaultActiveKey="1" flush style={{ marginTop: '10px' }}>
                <Accordion.Item eventKey="0">
                  <Accordion.Header>Add ingredients</Accordion.Header>
                  <Accordion.Body>
                    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
                    incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis
                    nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
                    Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu
                    fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
                    culpa qui officia deserunt mollit anim id est laborum.
                  </Accordion.Body>
                </Accordion.Item>
              </Accordion>
            </Form>
          </Col>
          <Col className="justify-content-md-center" md={{ span: 4, offset: 2 }}>
            <Card style={{ width: '26rem', textAlign: 'center', marginTop: '7%' }}>
              <Card.Img variant="top" src="https://s.tihlde.org/burger" />
              <Card.Body>
                <Card.Title>{this.recipeName}</Card.Title>
                <Card.Text> {this.description}</Card.Text>
                <Card.Text>{this.country}</Card.Text>
                <Card.Text>{this.category}</Card.Text>

                <Button variant="primary" data-toggle="collapse" data-target="#formBasicIngredient">
                  Gå videre
                </Button>
              </Card.Body>
            </Card>
          </Col>
        </Row>
      </Container>
    );
  }
}
