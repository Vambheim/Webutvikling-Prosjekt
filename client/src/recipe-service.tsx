import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type Recipe = {
  recipe_id: number;
  name: string;
  category: string;
  country: string;
};

export type RecipeDetailed = {
  recipe_id: number;
  name: string;
  category: string;
  country: string;
  ingriedients: Array<Ingredient>;
};

export type Step = {
  step_id: number;
  order_number: number;
  description: string;
  recipe_id: number;
};

export type RecipeIngredient = {
  ingredient_id: number;
  name: string;
  recipe_id: number;
  amount_per_person: number;
  measurement_unit: string;
};

export type Ingredient = {
  ingredient_id: number;
  name: string;
};

export type addIngredient = {
  name: string;
  amount: number;
  measurement_unit: string;
};

export type addStep = {
  description: string;
  order_number: number;
};

class RecipeService {
  /**
   * Get recipe with given recipe_id.
   */
  get(recipe_id: number) {
    return axios.get<Recipe>('/recipes/' + recipe_id).then((response) => response.data);
  }

  /**
   * Get all recipes.
   */
  getAll() {
    return axios.get<Recipe[]>('/recipes').then((response) => response.data);
  }

  /**
   * Create new recipe.
   *
   * Resolves the newly created task id.
   */
  createRecipe(name: string, category: string, country: string) {
    return axios
      .post<{ recipe_id: number }>('/recipes', {
        name: name,
        category: category,
        country: country,
      })
      .then((response) => response.data.recipe_id);
  }

  /**
   * Updates given recipe
   */
  update(recipe: Recipe) {
    return axios.put('/recipes', recipe).then((response) => response.data);
  }

  /**
   * Delete recipe with given id
   */
  delete(recipe_id: number) {
    return axios.delete('/recipes/' + recipe_id).then((response) => response.data);
  }

  /**
   * Get ingredients with given recipe_id
   */
  getRecipeIngredients(recipe_id: number) {
    return axios
      .get<RecipeIngredient[]>('/recipes/' + recipe_id + '/ingredients')
      .then((response) => response.data);
  }

  /**
   * Get all ingredients in the database
   */
  getAllIngredients() {
    return axios.get<Ingredient[]>('/ingredients').then((response) => response.data);
  }

  /**
   * Creates connection with recipe and ingredient
   */
  createRecipeIngredients(
    name: string,
    recipe_id: number,
    amount_per_person: number,
    measurement_unit: string
  ) {
    return axios
      .post('/recipes/ingredients', {
        name: name,
        recipe_id: recipe_id,
        amount_per_person: amount_per_person,
        measurement_unit: measurement_unit,
      })
      .then((response) => response.data);
  }

  /**
   * Updates given ingredient in given recipe
   */
  updateRecipeIngredient(
    amount_per_person: number,
    measurement_unit: string,
    recipe_id: number,
    ingredient_id: number,
    name: string
  ) {
    return axios
      .put('/recipes/' + recipe_id + '/ingredients/' + ingredient_id, {
        amount_per_person: amount_per_person,
        measurement_unit: measurement_unit,
        name: name,
      })
      .then((response) => response.data);
  }

  /**
   * Get all steps with given recipe_id
   */
  getSteps(recipe_id: number) {
    return axios.get<Step[]>('/recipes/' + recipe_id + '/steps').then((response) => response.data);
  }

  /**
   * Creates new step to given recipe
   */
  createStep(order_number: number, description: string, recipe_id: number) {
    return axios
      .post('/steps', {
        order_number: order_number,
        description: description,
        recipe_id: recipe_id,
      })
      .then((response) => response.data);
  }

  /**
   * Updates given step in given recipe
   */
  updateStep(recipe_id: number, step_id: number, order_number: number, description: string) {
    return axios
      .put('recipes/' + recipe_id + '/steps/' + step_id, {
        order_number: order_number,
        description: description,
      })
      .then((response) => response.data);
  }

  /**
   * Get recipes filtered by country and category
   */
  getFilterByCountryAndCategory(country: string, category: string) {
    return axios
      .get<Recipe[]>('/countryandcategoryfilter/' + country + '/' + category)
      .then((response) => response.data);
  }

  /**
   * Get recipes filtered by category
   */
  getFilterByCategory(category: string) {
    return axios.get<Recipe[]>('/categoryfilter/' + category).then((response) => response.data);
  }

  /**
   * Get recipes filtered by country
   */
  getFilterByCountry(country: string) {
    return axios.get<Recipe[]>('/countryfilter/' + country).then((response) => response.data);
  }

  /**
   * Get recipes filtered by one ingredient
   */
  getFilterByOneIngredient(ingredient1: string) {
    return axios
      .get<Recipe[]>('/oneingredientfilter/' + ingredient1)
      .then((response) => response.data);
  }

  /**
   * Get recipes filtered by two ingredients
   */
  getFilterBy2Ingredients(ingredient1: string, ingredient2: string) {
    return axios
      .get<Recipe[]>('/twoingredientsfilter/' + ingredient1 + '/' + ingredient2)
      .then((response) => response.data);
  }

  /**
   * Get recipes filtered by three ingredients
   */
  getFilterBy3Ingredients(ingredient1: string, ingredient2: string, ingredient3: string) {
    return axios
      .get<Recipe[]>(
        '/threeingredientsfilter/' + ingredient1 + '/' + ingredient2 + '/' + ingredient3
      )
      .then((response) => response.data);
  }

  /**
   * Get recommended recipes to given recipe
   */
  getRecommendedRecipes(recipe_id: number, category: string, country: string) {
    return axios
      .get<Recipe[]>('/recipes/' + recipe_id + '/recommended/' + category + '/' + country)
      .then((response) => response.data);
  }

  /**
   * Likes a given recipe from given user
   */
  likeRecipe(user_id: number, recipe_id: number) {
    return axios
      .post('/recipes/like', { user_id: user_id, recipe_id: recipe_id })
      .then((response) => response.data);
  }

  /**
   * Get liked recipes
   */
  getLikedRecipes(user_id: number) {
    return axios.get<Recipe[]>('/likedRecipes/' + user_id).then((response) => response.data);
  }

  /**
   * Posts recipes from spoonacular
   */
  PostSpoonacularRecipes(recipes: Array<RecipeDetailed>) {
    return axios
      .post<Array<RecipeDetailed>>('/spoonacular/recipes/', { recipes })
      .then((response) => response.data);
  }

  /**
   * Posts ingredients from spoonacular
   */
  PostSpoonacularIngriedents(ingridients: Array<RecipeIngredient>) {
    return axios
      .post<Array<RecipeIngredient>>('/spoonacular/ingridients/', { ingridients })
      .then((response) => response.data);
  }

  /**
   * Posts connection with ingredient and recipe from spoonacular
   */
  PostSpoonacularRecipeIngriedents(data: Array<RecipeDetailed>) {
    var ingridients = [];

    for (let i = 0; i < data.length;) {
      ingridients.push(data[i].ingriedients);
      i++;
    }
    ingridients = ingridients.flat();

    return axios
      .post<Array<RecipeIngredient>>('/spoonacular/ingridients-recipes/', { ingridients })
      .then((response) => response.data);
  }

  /**
   * Posts steps from spoonacular
   */
  PostSpoonacularSteps(steps: Array<Step>) {
    return axios
      .post<Array<Step>>('/spoonacular/steps/', { steps })
      .then((response) => response.data);
  }
}

const recipeService = new RecipeService();
export default recipeService;
