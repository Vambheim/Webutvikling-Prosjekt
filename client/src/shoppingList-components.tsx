import * as React from 'react';
import { Component } from 'react-simplified';
import { Alert, Column } from './widgets';
import shoppingListService, { ShoppingListInfo } from './shoppingList-service';
import { loggedIn, currentUser } from './user-components';
import { createHashHistory } from 'history';
import { Container, Card, Row, Form, Button, Col } from 'react-bootstrap';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a recipe

export class ShoppingList extends Component {
  shopping_list: ShoppingListInfo[] = [];

  render() {
    return (
      // <Container style={{ backgroundColor: 'yellow' }}>
      //   <Card style={{ backgroundColor: 'green' }}>
      //     <Card.Title>Shopping List</Card.Title>

      //     {this.shopping_list.map((list) => (
      //       <Row key={list.shopping_list_id}>
      //         <Column width={3}>
      //           {list.amount + ' ' + list.measurement_unit + ' ' + list.name}
      //         </Column>
      //         <Column width={1}>
      //           <Button
      //             variant="success"
      //             onClick={() => this.deleteOne(list.shopping_list_id, list.name)}
      //           >
      //             &#128465;
      //           </Button>
      //         </Column>
      //       </Row>
      //     ))}

      //     <Button variant="danger" onClick={() => this.deleteAll()}>
      //       Delete
      //     </Button>
      //   </Card>
      // </Container>

      <Card
        style={{
          // border: '5px',
          // borderStyle: ' solid',
          padding: '15px',
          textAlign: 'center',
          // marginLeft: 'auto',
          // marginRight: 'auto',
        }}
      >
        <Card.Title>Shopping list</Card.Title>

        <Row
          style={{
            marginLeft: '7%',
            width: '90%',
          }}
        >
          <Col>
            {this.shopping_list.map((list) => (
              <Row key={list.shopping_list_id}>
                <Col
                  style={{
                    textAlign: 'right',
                  }}
                >
                  {list.amount + ' ' + list.measurement_unit + ' ' + list.name}
                </Col>
                <Col>
                  <Button
                    variant="outline-success"
                    onClick={() => this.deleteOne(list.shopping_list_id, list.name)}
                    style={{
                      width: '5rem',
                      marginLeft: '0px',
                      marginRight: '100%',
                      marginBottom: '10px',
                    }}
                  >
                    &#128465;
                  </Button>
                </Col>
              </Row>
            ))}
          </Col>
        </Row>

        <Row>
          <Button
            variant="danger"
            onClick={() => this.deleteAll()}
            style={{
              width: '15rem',
              marginLeft: 'auto',
              marginRight: 'auto',
              marginBottom: '10px',
            }}
          >
            Delete all
          </Button>
        </Row>
      </Card>
    );
  }

  mounted() {
    if (!loggedIn) {
      Alert.info('Log in to use this feature');
    } else {
      shoppingListService
        .getShoppingList(currentUser.user_id)
        .then((list) => (this.shopping_list = list))
        .catch((error) => Alert.danger('Error getting shopping list ' + error.message));
    }
  }

  deleteOne(list_id: number, name: string) {
    if (confirm('Do you want to remove ' + name + ' from the shopping list?')) {
      console.log(list_id);
      shoppingListService
        .deleteItemShoppingList(list_id)
        .then(() => Alert.success('Item deleted'))
        .then(() => this.mounted()) // refreshes til items in shopping list
        .catch((error) => Alert.danger('Error deleting item in shopping list: ' + error.message));
    } else {
      console.log('Canceled');
    }
  }

  deleteAll() {
    if (!loggedIn) {
      Alert.info('Please log in');
    } else {
      if (this.shopping_list.length > 0) {
        if (confirm('Do you want to remove all items from shopping list?')) {
          shoppingListService
            .deleteShoppingList(currentUser.user_id)
            .then(() => Alert.info('Shopping list was successfully deleted'))
            .then(() => this.mounted())
            .catch((error) => Alert.danger('Error deleting shopping list ' + error.message));
        } else {
          console.log('Canceled');
        }
      } else {
        Alert.info('No items to delete');
      }
    }
  }
}
