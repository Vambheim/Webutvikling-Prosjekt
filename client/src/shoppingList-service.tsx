import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type ShoppingListInfo = {
  shopping_list_id: number;
  recipe_id: number;
  ingredient_id: number;
  name: string;
  amount: number;
  measurement_unit: string;
};

class ShoppingListService {
  /**
   * Get shopping list for given user
   */
  getShoppingList(user_id: number) {
    return axios
      .get<ShoppingListInfo[]>('/shoppinglist/' + user_id)
      .then((response) => response.data);
  }

  /**
   * Add to shopping list for given user
   */
  addToShoppingList(
    recipe_id: number,
    ingredient_id: number,
    user_id: number,
    amount: number,
    measurement_unit: string
  ) {
    return axios
      .post('/shoppinglist', {
        recipe_id: recipe_id,
        ingredient_id: ingredient_id,
        user_id: user_id,
        amount: amount,
        measurement_unit: measurement_unit,
      })
      .then((response) => response.data);
  }

  /**
   * Delete whole shopping list for given user
   */
  deleteShoppingList(user_id: number) {
    return axios.delete('/shoppinglist/' + user_id).then((response) => response.data);
  }

  /**
   * Delete one item in shopping list
   */
  deleteItemShoppingList(shopping_list_id: number) {
    return axios.delete('/shoppinglistitem/' + shopping_list_id).then((response) => response.data);
  }
}

const shoppingListService = new ShoppingListService();
export default shoppingListService;
