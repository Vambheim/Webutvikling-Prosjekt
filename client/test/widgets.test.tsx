import * as React from 'react';
import { shallow } from 'enzyme';
import { Column, Alert } from '../src/widgets';
import { Component } from 'react-simplified';

describe('Column widget tests', () => {
  test('Draws correctly', () => {
    const wrapper = shallow(<Column>test</Column>);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('New alert test', () => {
  test('No alerts initially', () => {
    const wrapper = shallow(<Alert />);

    expect(wrapper.matchesElement(<div></div>)).toEqual(true);
  });

  test('Show alert message', (done) => {
    const wrapper = shallow(<Alert />);

    Alert.danger('test');

    // Wait for events to complete
    setTimeout(() => {
      expect(
        wrapper.matchesElement(
          <div>
            <div>
              test
              <button />
            </div>
          </div>
        )
      ).toEqual(true);

      done();
    });
  });
  test('Alert.danger draws correctly with classname', () => {
    //@ts-ignore
    const wrapper = shallow(<Alert.danger>test</Alert.danger>);

    expect(wrapper).toMatchSnapshot();
  });

  test('Alert.success draws correctly with classname', () => {
    //@ts-ignore
    const wrapper = shallow(<Alert.success>test</Alert.success>);

    expect(wrapper).toMatchSnapshot();
  });

  test('Button.Primary draws correctly with classname and style', () => {
    //@ts-ignore
    const wrapper = shallow(<Alert.warning>test</Alert.warning>);
    //console.log(wrapper.debug());

    expect(wrapper).toMatchSnapshot();
  });

  test('Button.info draws correctly with classname', () => {
    //@ts-ignore
    const wrapper = shallow(<Alert.info>test</Alert.info>);
    //console.log(wrapper.debug());

    expect(wrapper).toMatchSnapshot();
  });

  test('Show alert message', (done) => {
    const wrapper = shallow(<Alert />);

    Alert.danger('test');

    // Wait for events to complete
    setTimeout(() => {
      expect(
        wrapper.containsMatchingElement(
          <div>
            <div>
              test
              <button />
            </div>
          </div>
        )
      ).toEqual(true);

      done();
    });
  });

  test('Alert draws correctly with classname', () => {
    const wrapper = shallow(<Alert>test</Alert>);

    expect(wrapper).toMatchSnapshot();
  });

  test('Alert.danger draws correctly with classname', () => {
    //@ts-ignore
    const wrapper = shallow(<Alert.danger>test</Alert.danger>);

    expect(wrapper).toMatchSnapshot();
  });
});
