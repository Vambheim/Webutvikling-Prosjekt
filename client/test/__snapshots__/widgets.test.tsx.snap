// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column widget tests Draws correctly 1`] = `
<div
  className="col"
>
  <div
    className="float-start"
  >
    test
  </div>
</div>
`;

exports[`New alert test Alert draws correctly with classname 1`] = `<div />`;

exports[`New alert test Alert.danger draws correctly with classname 1`] = `""`;

exports[`New alert test Alert.danger draws correctly with classname 2`] = `""`;

exports[`New alert test Alert.success draws correctly with classname 1`] = `""`;

exports[`New alert test Button.Primary draws correctly with classname and style 1`] = `""`;

exports[`New alert test Button.info draws correctly with classname 1`] = `""`;

exports[`New alert test Button.info draws correctly with classname and style 1`] = `""`;
